tempesta_fw/apm.c

```diff
@@ -372,6 +373,111 @@ tfw_stats_update(TfwPcntRanges *rng, unsigned int r_time)
 	return;
 }
 
+/* Time granularity for HTTP codes accounting during health monitoring. */
+#define HM_FREQ		10
+
+/*
+ * Structure for health monitor settings.
+ *
+ * @list	- entry in list of all health monitors;
+ * @name	- health monitor's name;
+ * @url		- url for requests which will be used in health monitoring;
+ * @urlsz	- length of @url string (without terminating zero);
+ * @req		- full test request health monitoring;
+ * @reqsz	- length of @req string (without terminating zero);
+ * @crc32	- crc32 value for verification of response body checksum;
+ * @codes	- pointer to HTTP response codes bitmap (signals that
+ *		  backend server alive);
+ * @tmt		- timeout between health monitoring requests;
+ * @auto_crc	- flag for enabling of crc32 generation from first response;
+ */
+typedef struct {
+	struct list_head	list;
+	char			*name;
+	char			*req;
+	unsigned long		reqsz;
+	char			*url;
+	int			urlsz;
+	u32			crc32;
+	long			*codes;
+	unsigned short		tmt;
+	bool			auto_crc:1;
+} TfwApmHM;
+
+/*
+ * Structure for monitoring settings of particular HTTP code.
+ *
+ * @list	- entry in list of all monitored codes;
+ * @tframe	- Time frame in seconds for @code accounting;
+ * @limit	- allowed quantity of responses with @code in a @tframe period;
+ * @code	- HTTP code; also wildcarded code values (of type 4*, 5* etc.)
+ *		  are allowed during configuration, so in this field they will
+ *		  have form of single-digit number (e.g. 4 or 5 respectively);
+ */
+typedef struct {
+	struct list_head	list;
+	unsigned short		tframe;
+	unsigned short		limit;
+	int			code;
+} TfwApmHMCfg;
+
+/*
+ * History accounting record.
+ *
+ * @ts		- part of timeframe in granularity of HM_FREQ;
+ * @resp	- amount responses counted in @ts interval;
+ */
+typedef struct {
+	unsigned long		ts;
+	unsigned int		resp;
+} TfwApmHMHistory;
+
+/*
+ * Accounting entry for particular HTTP code.
+ *
+ * @history	- ring buffer of history records ;
+ * @hmcfg	- pointer to structure with settings for particular HTTP code;
+ * @rsum	- current amount of responses from @history ring buffer;
+ * @lock	- spinlock for synchronized work with @history buffer;
+ */
+typedef struct {
+	TfwApmHMHistory		history[HM_FREQ];
+	TfwApmHMCfg		*hmcfg;
+	unsigned int		rsum;
+	spinlock_t		lock;
+} TfwApmHMStats;
+
+/*
+ * Controller for whole health monitoring of backend server.
+ *
+ * @hm		- pointer to settings for specific health monitor;
+ * @hmstats	- pointer to array of stat entries for all monitored HTTP codes;
+ * @rcount	- current count of health monitoring requests (in @hm->tmt);
+ * @jtmstamp	- time in jiffies of last @timer call (for procfs);
+ * @timer	- timer for sending health monitoring request;
+ * @rearm	- flag for gracefull stopping of @timer;
+ */
+typedef struct {
+	TfwApmHM		*hm;
```

Comment:
Seems at least `hmstats`, `hm` are server group-wide options and should be stored there. We can also use one timer per server group: send health check message to all server in group. What do you think? not the comment below.

Reply:
We cannot make `hmstats` group-wide - since this will break the essence of the task (server-wide HTTP health statistics and server-wide suspend/alive decisions).
As for `hm` - yes, theoretically we can use it group-wide, but for now I'd prefer leave it server-wide for three reasons:
1. Changing current implementation (originally grounded on a server-wide approach) will require significant code changes and re-testing.
2. In current approach all health monitoring functionality is concentrated in APM module and the there is no need to spreading it between other modules (e.g `sock_srv` etc.)
3. Current solution is more flexible and give more granularity in HM management if it will be needed in future, and I do not see significant advantages in group-wide approach except of saving memory on `hm` pointers for each server.